rllib/agents/pg/pg.py

```diff
@@ -32,3 +86,228 @@ def get_default_config(cls) -> TrainerConfigDict:
+        weights_all_trained_policy = ray.get(
+            [w.get_weights.remote() for w in self._training_workers]
+        )
+        for i in range(len(weights_all_trained_policy)):
+            for j in range(i, len(weights_all_trained_policy)):
+                check_weights(
+                    weights_all_trained_policy[i],
+                    weights_all_trained_policy[j],
+                    atol=5e-2,
+                )
+
+    def get_shard_idxs(self, train_batch: SampleBatch):
+        def get_shard_idxs_helper(_train_batch: SampleBatch):
+            train_batch_shard_idx = []
+            start = 0
+            batch_size = len(_train_batch) // self._num_train_workers
+            for i in range(self._num_train_workers):
+                train_batch_shard_idx.append(
+                    (start, min(start + batch_size, len(_train_batch)))
+                )
+                start = start + batch_size
+            return train_batch_shard_idx
+
+        if isinstance(train_batch, SampleBatch):
+            return get_shard_idxs_helper(train_batch)
+        elif isinstance(train_batch, MultiAgentBatch):
+            idxs = {}
+            for name, batch in train_batch.policy_batches.items():
+                idxs[name] = get_shard_idxs_helper(batch)
+            ret = [{} for _ in range(self._num_train_workers)]
+            for name, idx in idxs.items():
+                for i, v in enumerate(idx):
+                    ret[i][name] = v
+            return ret
+
+
+class PGTrainWorkerTorch:
+    def __init__(self, config):
+        self.config = copy.deepcopy(config)
+
+    def start(self, rank, world_size):
+        if self.config["seed"]:
+            update_global_seed_if_necessary("torch", self.config["seed"])
+        self.config["training_worker_rank"] = rank
+        self.config["number_of_training_workers"] = world_size
+        if not self.config.get("observation_space") or not self.config.get(
+            "action_space"
+        ):
+            spaces = self.config["spaces_inferred_from_sampling_workers"]
+        else:
+            spaces = None
+
+        self._local_worker = Trainer.make_worker(
+            cls=RolloutWorker,
+            env_creator=lambda _: None,
+            policy_cls=self.config["default_policy_cls"],
+            worker_index=0,
+            num_workers=self.config["number_of_training_workers"],
+            config=self.config,
+            spaces=spaces,
+            validate_env=None,
+        )
+        if not (world_size > 1):
+            return
+        for pid in self._local_worker.policies_to_train:
+            model = self._local_worker.policy_map[pid].model
+            if isinstance(model, DDPTorchModelV2Wrapper) or not isinstance(
+                model, TorchModelV2
+            ):
+                continue
+            if torch.cuda.is_available():
+                model = DDPTorchModelV2Wrapper(
+                    model, device_ids=[rank], output_device=rank
```

Comment:
Not sure if you're using (multi-node) GPU yet but this should be local rank (as opposed to global rank)